src/os/linux/mod.rs

```diff
@@ -0,0 +1,168 @@
+                })
+            })
+            .ok_or(
+                // Because if it found an entry then it must've failed while parsing it.
+                // Otherwise we wouldn't be here, as it would've returned Some(...)
+                if some_entry_found {
+                    DetectEditorError::FreeDesktopEntryParseError
+                } else {
+                    DetectEditorError::FreeDesktopEntryNotFound
+                },
+            )?)
+    }
+
+    pub fn open_file(&self, path: impl AsRef<Path>) -> Result<(), OpenFileError> {
+        let path = path.as_ref();
+
+        let maybe_icon = self.icon.as_ref().map(|icon_str| icon_str.as_os_str());
+
+        // Parse the xdg command field with all the needed data
+        let command_parts = xdg::parse_command(
+            &self.exec_command,
+            path.as_os_str(),
+            maybe_icon,
+            Some(&self.xdg_entry_path),
+        );
+
+        if !command_parts.is_empty() {
+            // If command_parts has at least one element this works. If it has a single
+            // element, &command_parts[1..] should be an empty slice (&[]) and bossy
+            // `with_args` does not add any argument on that case, although the docs
+            // do not make it obvious.
+            bossy::Command::impure(&command_parts[0])
+                .with_args(&command_parts[1..])
+                .run_and_detach()
+                .map_err(OpenFileError::LaunchFailed)?;
+            Ok(())
+        } else {
+            Err(OpenFileError::CommandParsingFailed)
+        }
+    }
+}
+
+pub fn open_file_with(
+    application: impl AsRef<OsStr>,
+    path: impl AsRef<OsStr>,
+) -> bossy::Result<()> {
+    let app_str = application.as_ref();
+    let path_str = path.as_ref();
+
+    let command_parts = xdg::get_xdg_data_dirs()
+        .iter()
+        .find_map(|dir| {
+            let dir = dir.join("applications");
+            let (entry, entry_path) = xdg::find_entry_by_app_name(&dir, &app_str)?;
+
+            let command_parts =
+                entry
+                    .section("Desktop Entry")
+                    .attr("Exec")
+                    .and_then(|str_entry| {
+                        let osstring_entry: OsString = str_entry.into();
+                        Some(xdg::parse_command(
+                            &osstring_entry,
+                            path_str,
+                            entry
+                                .section("Desktop Entry")
+                                .attr("Icon")
+                                .map(|s| s.as_ref()),
+                            Some(&entry_path),
+                        ))
+                    })?;
+            // This could go outside, but we'd better have a proper error for it then
+            if !command_parts.is_empty() {
+                Some(command_parts) // This guarantees that command_parts has at least one element
+            } else {
+                None
```

Comment:
I think we should check on every directory (until we find something). This is because applications can be installed in a variety of different ways, and depending on the installation the `.desktop` file may end up on a different directory. In my system, for instance, Flatpak has its own directory for its desktop files, that it adds to `$XDG_DATA_DIRS` variable. I don't think we can ever only check a single dir.

Reply:
Agreed; sorry if I phrased it oddly, but I was asking about the fact that this logic continues searching if a match is found but `command_parts` is empty. `commands_parts` being empty would mean the desktop entry is malformed, right? (you surely know the spec better than I do)

Reply:
I get it now, I remember the spec says the Exec field is not always required, it is only required in case the entry says it is DBus "activatable", recommending to always set the Exec field anyway, for compatibility. Now, I don't really know a lot about DBus, and when I read that for the first time, I sort of ignored it, as I've never ever launched any application through DBus. Everything I've ever launched via a terminal, I've launched by typing the name of its executable, like usual.

This means lacking a Exec entry does not mean the entry is malformed (it may be the case, but not always). However, I think it may be better to fail in case an Exec field is not found. That would be compliant to the 1.0 version of the spec, which still does not require an Exec field. If there is no Exec field we just can't proceed, unless there is another file with that same filename in another xdg dir, for some reason.

If we want to support the 1.1 version, we should have to implement this DBus activation, but I don't know if it will ever be used. Even if there is an application that supports being launched with DBus, it likely still would have an Exec field. An application that has no Exec field, probably does not want to be launched from a "dirty" environment (such as a terminal with a bunch of random env variables), and only want to be launched from a "clean" environment, such as being launched from the system tray.

Reply:
Makes sense, and sounds good! Thanks for being so thorough. I don't know much about dbus either, but I share your optimism that we probably don't need to account for it.